File: web/src/storage/model/config/partition.ts
/*
 * Copyright (c) [2024] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

import { config } from "~/api/storage/types";
import * as checks from "~/api/storage/types/checks";
import { Size, generate as generateSize } from "~/storage/model/config/size";
import { generateName, generateFilesystem, generateSnapshots } from "~/storage/model/config/common";

export type Partition = {
  index?: number;
  name?: string;
  alias?: string;
  delete?: boolean;
  deleteIfNeeded?: boolean;
  resize?: boolean;
  resizeIfNeeded?: boolean;
  filesystem?: string;
  mountPath?: string;
  snapshots?: boolean;
  size?: Size;
};

export type PartitionConfig =
  | config.RegularPartition
  | config.PartitionToDelete
  | config.PartitionToDeleteIfNeeded;

export function isPartitionConfig(
  partition: config.PartitionElement,
): partition is PartitionConfig {
  return (
    checks.isRegularPartition(partition) ||
    checks.isPartitionToDelete(partition) ||
    checks.isPartitionToDeleteIfNeeded(partition)
  );
}

type PartitionWithSizeConfig = config.RegularPartition | config.PartitionToDeleteIfNeeded;

function isPartitionWithSizeConfig(
  partition: config.PartitionElement,
): partition is PartitionWithSizeConfig {
  return checks.isRegularPartition(partition) || checks.isPartitionToDeleteIfNeeded(partition);
}

class PartitionGenerator {
  private partitionConfig: config.PartitionElement | undefined;
  private solvedPartitionConfig: PartitionConfig;

  constructor(
    partitionConfig: config.PartitionElement | undefined,
    solvedPartitionConfig: PartitionConfig,
  ) {
    this.partitionConfig = partitionConfig;
    this.solvedPartitionConfig = solvedPartitionConfig;
  }

  generate(): Partition {
    if (checks.isRegularPartition(this.solvedPartitionConfig)) {
      return this.fromRegularPartition(this.solvedPartitionConfig);
    } else if (checks.isPartitionToDelete(this.solvedPartitionConfig)) {
      return this.fromPartitionToDelete(this.solvedPartitionConfig);
    } else if (checks.isPartitionToDeleteIfNeeded(this.solvedPartitionConfig)) {
      return this.fromPartitionToDeleteIfNeeded(this.solvedPartitionConfig);
    }
  }

  private fromRegularPartition(solvedPartitionConfig: config.RegularPartition): Partition {
    return {
      index: this.solvedPartitionConfig.index,
      name: generateName(solvedPartitionConfig),
      alias: solvedPartitionConfig.alias,
      resize: this.generateResize(),
      resizeIfNeeded: this.generateResizeIfNeeded(),
      filesystem: generateFilesystem(solvedPartitionConfig),
      mountPath: solvedPartitionConfig.filesystem?.path,
      snapshots: generateSnapshots(solvedPartitionConfig),
      size: this.generateSize(),
    };
  }

  private fromPartitionToDelete(solvedPartitionConfig: config.PartitionToDelete): Partition {
    return {
      index: this.solvedPartitionConfig.index,
      name: generateName(solvedPartitionConfig),
      delete: true,
    };
  }

  private fromPartitionToDeleteIfNeeded(
    solvedPartitionConfig: config.PartitionToDeleteIfNeeded,
  ): Partition {
    return {
      index: this.solvedPartitionConfig.index,
      name: generateName(solvedPartitionConfig),
      deleteIfNeeded: true,
      resize: this.generateResize(),
      resizeIfNeeded: this.generateResizeIfNeeded(),
      size: this.generateSize(),
    };
  }

  private generateSize(): Size | undefined {
    if (!isPartitionWithSizeConfig(this.solvedPartitionConfig)) return;

    return generateSize(this.partitionConfig, this.solvedPartitionConfig);
  }

  // TODO: return false if the size is equal to the size of the system device.
  private generateResize(): boolean | undefined {
    if (this.solvedPartitionConfig.search === undefined) return;

    const size = this.generateSize();
    return size !== undefined && !size.auto && size.min !== undefined && size.min === size.max;
  }

  private generateResizeIfNeeded(): boolean | undefined {
    if (this.solvedPartitionConfig.search === undefined) return;

    const size = this.generateSize();
    return size !== undefined && !size.auto && size.min !== size.max;
  }
}

export function generate(
  partitionConfig: config.PartitionElement | undefined,
  solvedPartitionConfig: PartitionConfig,
): Partition {
  const generator = new PartitionGenerator(partitionConfig, solvedPartitionConfig);
  return generator.generate();
}

File: service/lib/agama/storage/config_conversions/to_model_conversions/partition.rb
# frozen_string_literal: true

# Copyright (c) [2024] SUSE LLC
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of version 2 of the GNU General Public License as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, contact SUSE LLC.
#
# To contact SUSE LLC about this file by physical or electronic mail, you may
# find current contact information at www.suse.com.

require "agama/storage/config_conversions/to_model_conversions/base"
require "agama/storage/config_conversions/to_model_conversions/with_filesystem"
require "agama/storage/config_conversions/to_model_conversions/with_size"
require "agama/storage/configs/partition"

module Agama
  module Storage
    module ConfigConversions
      module ToModelConversions
        # Partition conversion to model according to the JSON schema.
        class Partition < Base
          include WithFilesystem
          include WithSize

          # @see Base
          def self.config_type
            Configs::Partition
          end

        private

          # @see Base#conversions
          def conversions
            {
              name:           config.found_device&.name,
              alias:          config.alias,
              id:             config.id&.to_s,
              mountPath:      config.filesystem&.path,
              filesystem:     convert_filesystem,
              size:           convert_size,
              delete:         config.delete?,
              deleteIfNeeded: config.delete_if_needed?,
              resize:         convert_resize,
              resizeIfNeeded: convert_resize_if_needed
            }
          end

          # @return [Booelan]
          def convert_resize
            size = config.size

            !size.nil? && !size.default? && size.min == size.max
          end

          # @return [Booelan]
          def convert_resize_if_needed
            size = config.size

            !size.nil? && !size.default? && size.min != size.max
          end
        end
      end
    end
  end
end
